Handle concatenation nodes in bash command parser for exec policy

ivanmurashko opened the pull request:
The bash command parser in exec_policy was failing to parse commands with concatenated flag-value patterns like `-g"*.py"` (no space between flag and quoted value). This caused policy rules like `prefix_rule(pattern=["rg"])` to not match commands such as `rg -n "foo" -g"*.py"`.

When tree-sitter-bash parses `-g"*.py"`, it creates a "concatenation" node containing a word (`-g`) and a string (`"*.py"`). The parser previously rejected any node type not in the ALLOWED_KINDS list, causing the entire command parsing to fail and fall back to matching against the wrapped `bash -lc` command instead of the inner command.

This change:
- Adds "concatenation" to ALLOWED_KINDS in try_parse_word_only_commands_sequence
- Adds handling for concatenation nodes in parse_plain_command_from_node that recursively extracts and joins word/string/raw_string children
- Adds test cases for concatenated flag patterns with double and single quotes

Fixes #8394

ivanmurashko committed:
Handle concatenation nodes in bash command parser for exec policy

The bash command parser in exec_policy was failing to parse commands with
concatenated flag-value patterns like `-g"*.py"` (no space between flag and
quoted value). This caused policy rules like `prefix_rule(pattern=["rg"])`
to not match commands such as `rg -n "foo" -g"*.py"`.

When tree-sitter-bash parses `-g"*.py"`, it creates a "concatenation" node
containing a word (`-g`) and a string (`"*.py"`). The parser previously
rejected any node type not in the ALLOWED_KINDS list, causing the entire
command parsing to fail and fall back to matching against the wrapped
`bash -lc` command instead of the inner command.

This change:
- Adds "concatenation" to ALLOWED_KINDS in try_parse_word_only_commands_sequence
- Adds handling for concatenation nodes in parse_plain_command_from_node that
  recursively extracts and joins word/string/raw_string children
- Adds test cases for concatenated flag patterns with double and single quotes

Fixes #8394

bolinfest reviewed (commented):
@ivanmurashko once again, really nice work here! Two small things and then I think we're good to go!

ivanmurashko committed:
Fail explicitly on empty concatenation in bash parser

Change empty concatenation handling from silently skipping to explicitly
returning None. This makes the error handling consistent with other
validation failures in the function and prevents silently accepting
malformed input.

An empty concatenation indicates either a bug in the iteration logic or
unexpected AST structure from tree-sitter-bash. Returning None ensures
these edge cases are caught rather than being silently ignored.

ivanmurashko committed:
Add tests for rejecting unsafe constructs in concatenations

Add test cases to verify that the bash parser correctly rejects
environment variable substitution and command substitution within
concatenated arguments.

These tests ensure that patterns like:
- rg -g"$VAR" (variable substitution)
- rg -g"${VAR}" (brace expansion)
- rg -g"$(pwd)" (command substitution)

are properly rejected by the parser. While the implementation already
handles these cases correctly (tree-sitter creates variable/command_substitution
nodes instead of string_content, causing validation to fail), these tests
explicitly document and verify the rejection behavior.

bolinfest reviewed (approved):
Thanks for the fast reply and very nice work!
